### PR TITLE
Added support for the official Swift 3.0

### PR DESCRIPTION
**Pull request by ilap**

### CODE REVIEW

File: Sources/CLI.swift [hunk 1]
```diff
@@ -72,7 +72,7 @@ public class CLI {
         - Parameter commandName: the name of the new chainable command
         - Returns: a new chainable command for immediate chaining
     */
-    public class func registerChainableCommand(commandName commandName: String) -> ChainableCommand {
+    public class func registerChainableCommand(_ commandName: String) -> ChainableCommand {
```
Comment (jakeheis) [inline Sources/CLI.swift]: Swift 3 equivalent is `registerChainableCommand(commandName: String)`

File: Sources/CLI.swift [hunk 2]
```diff
@@ -128,7 +128,7 @@ public class CLI {
     
     // MARK: - Privates
     
-    class private func routeCommand(arguments arguments: RawArguments) throws -> CommandType {
+    class private func routeCommand(_ arguments: RawArguments) throws -> CommandType {
```
Comment (jakeheis) [inline Sources/CLI.swift]: `routeCommand(arguments: RawArguments)`

File: Sources/CommandArguments.swift [hunk 3]
```diff
@@ -82,15 +82,15 @@ public class CommandArguments {
...
         guard rawArguments.unclassifiedArguments().count == 0  else {
-            throw CLIError.Error("Expected no arguments, got \(rawArguments.unclassifiedArguments().count).")
+            throw CLIError.error("Expected no arguments, got \(rawArguments.unclassifiedArguments().count).")
         }
     
         return CommandArguments()
     }
     
-    private class func errorMessage(expectedCount expectedCount: Int, givenCount: Int) -> String {
+    private class func errorMessage(_ expectedCount: Int, givenCount: Int) -> String {
```
Comment (jakeheis) [inline Sources/CommandArguments.swift]: `errorMessage(expectedCount: Int, givenCount: Int)`

File: Sources/CommandMessageGenerator.swift [hunk 4]
```diff
@@ -8,7 +8,7 @@
 
 class CommandMessageGenerator {
     
-    class func generateUsageStatement(command command: CommandType, options: Options?) -> String {
+    class func generateUsageStatement(_ command: CommandType, options: Options?) -> String {
```
Comment (jakeheis) [inline Sources/CommandMessageGenerator.swift]: `generateUsageStatement(command: CommandType, options: Options?)`

File: Sources/Extensions.swift [hunk 5]
```diff
@@ -8,7 +8,7 @@
 
 extension String {
     
-    func padFront(totalLength totalLength: Int) -> String {
+    func padFront(_ totalLength: Int) -> String {
```
Comment (jakeheis) [inline Sources/Extensions.swift]: `padFront(totalLength: Int)`

File: Sources/Input.swift [hunk 6]
```diff
@@ -23,7 +23,7 @@ public class Input {
         - Parameter message: message to be printed before accepting input (e.g. "Name: ")
         - Throws: Input.PipeUserInputOverlapError if the user piped data and this method was called
     */
-    public class func awaitInput(message message: String?) throws -> String {
+    public class func awaitInput(_ message: String?) throws -> String {
```
Comment (jakeheis) [inline Sources/Input.swift]: `awaitInput(message: String?)` and same for the rest of the methods in this file

File: Sources/Input.swift [hunk 7]
```diff
@@ -100,10 +100,10 @@ public class Input {
     
     class func checkForPipedData() {
         inputHandle.readabilityHandler = {(inputHandle) in
-            pipedData = String(data: inputHandle.availableData, encoding: NSUTF8StringEncoding)
+            pipedData = String(data: inputHandle.availableData, encoding: String.Encoding.utf8)
             inputHandle.readabilityHandler = nil
         }
-        NSProcessInfo.processInfo().arguments // For whatever reason, this triggers readabilityHandler for the pipe data
+        //FIXME: ProcessInfo.processInfo().arguments // For whatever reason, this triggers readabilityHandler for the pipe data
```
Comment (jakeheis) [inline Sources/Input.swift]: What has broken here?

File: Sources/RawArguments.swift [hunk 8]
```diff
@@ -45,41 +45,41 @@ public class RawArguments: CustomStringConvertible {
     
     init(arguments: [String]) {
         self.arguments = arguments
-        self.argumentClassifications = [RawArgumentType](count: arguments.count, repeatedValue: .Unclassified)
+        self.argumentClassifications = [RawArgumentType](repeating: .unclassified, count: arguments.count)
         
-        classifyArgument(index: 0, type: .AppName)
+        classifyArgument(0, type: RawArgumentType.appName)
```
Comment (jakeheis) [inline Sources/RawArguments.swift]: Emit the enum type when possible, `classifyArgument(0, type: .appName)`

File: Sources/RawArguments.swift [hunk 9]
```diff
@@ -45,41 +45,41 @@ public class RawArguments: CustomStringConvertible {
...
     }
     
-    public func classifyArgument(argument argument: String, type: RawArgumentType) {
-        if let index = arguments.indexOf(argument) {
-            classifyArgument(index: index, type: type)
+    public func classifyArgument(_ argument: String, type: RawArgumentType) {
```
Comment (jakeheis) [inline Sources/RawArguments.swift]: `public func classifyArgument(argument: String, type: RawArgumentType)`

File: Sources/RawArguments.swift [hunk 10]
```diff
@@ -45,41 +45,41 @@ public class RawArguments: CustomStringConvertible {
...
+        if let index = arguments.index(of: argument) {
+            classifyArgument(index, type: type)
         }
     }
     
-    private func classifyArgument(index index: Int, type: RawArgumentType) {
+    private func classifyArgument(_ index: Int, type: RawArgumentType) {
```
Comment (jakeheis) [inline Sources/RawArguments.swift]: `classifyArgument(index: Int, type: RawArgumentType)`

**Comment by jakeheis** [comment]
Thank you for your help with this! I left comments on the methods that appeared to have had their signatures incorrectly changed in their transition to Swift 3. If you could fix the remaining issues, I'll merge it in!

**Comment by ilap** [comment]
Hi,

I have fixed the wrong signatures, but in Swift 3.0 the singatures like
this func(_ param1: type, param2: type) can be resolved by func(1, param2:
10) instead of func(param1: 1, param2: 10).
This is recommended for the Swift 3.0 evolution and this can eliminate the
requirement for the label for the 1st parameter.

Cheers,

Pal

File: Sources/CLI.swift [hunk 11]
```diff
@@ -32,7 +32,7 @@ public class CLI {
         - Parameter version: version of the app, printed by the VersionCommand
         - Parameter description: description of the app, printed in the help message
     */
-    public class func setup(name name: String, version: String? = nil, description: String? = nil) {
+    public class func setup(_ name: String, version: String? = nil, description: String? = nil) {
```
Comment (jakeheis) [inline Sources/CLI.swift]: `setup(name: String, version: String? = nil, description: String? = nil)`

File: Sources/CommandArguments.swift [hunk 12]
```diff
@@ -82,15 +82,15 @@ public class CommandArguments {
         return commandArguments
     }
     
-    private class func handleEmptySignature(rawArguments rawArguments: RawArguments) throws -> CommandArguments {
+    private class func handleEmptySignature(_ rawArguments: RawArguments) throws -> CommandArguments {
```
Comment (jakeheis) [inline Sources/CommandArguments.swift]: `handleEmptySignature(rawArguments: RawArguments)`

**Comment by jakeheis** [comment]
With this, in what will likely become SwiftCLI 1.3, I'm trying to maintain the same interface with all the same function signatures. A future release will provide more Swift-3.0-y names. If you could clear up the remaining issues so the interface is the same, I'd much appreciate it.

**Comment by jakeheis** [comment]
Superseded by fc0697459818b9e86c88a98ef59c81f8d1d13543; thanks for your contribution!

**Pull request closed without merging**